ports-of-call/portable_arrays.hpp

```diff
@@ -26,73 +26,93 @@
 
 #include "portability.hpp"
 #include <algorithm>
+#include <array>
 #include <assert.h>
 #include <cstddef> // size_t
 #include <cstring> // memset()
+#include <functional>
+#include <numeric>
+#include <type_traits>
 #include <utility> // swap()
 
+// maximum number of dimensions
+constexpr std::size_t MAXDIM = 6;
+
+namespace detail {
+// convert `index_sequence` values to constant
+template <std::size_t I, std::size_t V>
+constexpr std::size_t to_const = V;
+
+// array type of dimensions/strides
+// multiply reduce array
+// NOTE: we can do product of variadic parameters `vars...` as
+// `arr_mul({vars...})`
+template <typename T, std::size_t N>
+PORTABLE_INLINE_FUNCTION auto arr_mul(const std::array<T, N> &a) {
+  auto r = T{1};
+  for (auto v : a)
+    r *= v;
+  return r;
+}
+PORTABLE_FORCEINLINE_FUNCTION
+decltype(auto) vp_prod() {
```

Comment:
is the `decltype` needed? Why not just an `auto` return value?

Reply:
eh, I'm returning a generic lambda and that's my habbit. But it's not necessary here (tho I don't think it's harmful either)

Reply:
I don't think the `decltype` adds anything and it's a bit harder to read but I won't push back too much on this.